### ISSUE TITLE
scripts: run.command blocked if command produces lots of output

### ISSUE THREAD

**Issue by maxpietsch** [MRtrix3/mrtrix3#927]

**Comment by maxpietsch** [comment]
`run.command ('cat /tmp/text | echo')` also fails to finish if /tmp/text is larger than about 100KB  (b7424ccf79cae3b5597bd8ccdd13e) The reason I am looking at this is that `population_template` using many subjects can create that much output which causes scripts that call it to freeze.

**Comment by maxpietsch** [comment]
[This](https://stackoverflow.com/questions/375427/non-blocking-read-on-a-subprocess-pipe-in-python/) suggest it can be done in a cross-platform compatible way using [queues](https://docs.python.org/3/library/queue.html#Queue.Queue.get) and a separate thread to do the reading.

**Comment by maxpietsch** [comment]
Ok, I think I got it to work on my system. Below code generates a 12MB text file and pipes it into a list without locking up. The trick is that the reader is in a separate thread (and defined as demon so that it is killed with the process x): 

    #!/usr/bin/env python
    import subprocess
    from threading import Thread
    import time
    
    from random import choice
    import string
    def gen_line(length=8, chars=string.letters + string.digits):
       return ''.join([choice(chars) for i in range(length)])
    
    print "generating text file"
    with open('/tmp/text','w') as f:
      for i in range(200000):
        f.write("%i: %s\n" % (i, gen_line(50,string.ascii_letters)))
    print "done"
    
    def reader(f,buffer):
       while True:
         line=f.readline()
         if line:
            buffer.append(line)
         else:
            print "reader has finished"
            break
    
    linebuffer=[]
    x=subprocess.Popen(['cat', '/tmp/text'],stdout=subprocess.PIPE)
    t=Thread(target=reader,args=(x.stdout,linebuffer))
    t.daemon=True
    t.start()
    
    # below is for near real time printing the output, could just use t.join() here
    while True:
      time.sleep(0.1)
      if len(linebuffer):
        print "line " + linebuffer.pop(0),
      else:
        if not t.isAlive() and len(linebuffer) == 0:
          print "finished"
          t.join()
          break
        print "waiting for output"
        time.sleep(1)

**Comment by Lestropie** [comment]
ec47f19 fixed the deadlock for single command invocations by using Python's `tempfile` module. I wasn't aware that the same deadlock could occur when piping between processes; in fact I'd have expected it *not* to. I was trying to avoid the dedicated-thread-for-reading solution since `run.command()` could have an arbitrary number of commands to run; plus there's managing the `-info` / `-debug` options. But maybe I don't get a choice any more... At least now it would only be dedicated threads for handling pipes, rather than all process output.

@maxpietsch Are you anywhere near having something that can be incorporated into `run.command()`? Or do I need to attack it from scratch?

**Comment by maxpietsch** [comment]
Donald and I had a good look at the issue. The problem is that my test is not testing what I thought. `cat file | echo` does not print the content of the file, `cat file | cat ` does (and works with `run.command`).  So we won't need to faff about with threads for reading or writing.

`run.command("cat /tmp/text | echo")` locks up because we try to write to a broken pipe. The same  happens for `run.command("base64 /dev/urandom | head -c 10000000")` where `head` finishes but `base64 /dev/urandom` never does. 

So to fix this we need a [check](https://docs.python.org/2/library/subprocess.html#replacing-shell-pipeline) for broken pipes. I think  fe0916429a3c6832ef should do that:


```
/home/mp14/mrtrix3/bin/runcommand_blocked 1 
base64: write error: Broken pipe
base64: write error
runcommand_blocked: output = run.command ('cat /tmp/text | cat')
Command:  cat /tmp/text | cat
len(output[0]):  10000000
Command:  cat /tmp/text | echo
runcommand_blocked: 
runcommand_blocked: [ERROR] Command failed: cat /tmp/text | echo (runcommand_blocked:27)
runcommand_blocked: Output of failed command:
cat: write error: Broken pipe
runcommand_blocked: Changing back to original directory (/home/mp14/mrtrix3)
```

@Lestropie  I'd suggest cherry-picking that commit into the tag0.3.16 branch if you're happy with it.

**Comment by Lestropie** [comment]
I can't remember where it ended up, but I did show a pipe example recently where an error in a command resulted in the other command receiving a broken pipe system signal and terminating via the *MRtrix3* custom handler. So even in the case of such malformed `run.command()` calls, it's only non-*MRtrix3* commands without signal handlers that will hang.

My problem with fe09164 (if I'm interpreting it correctly) is that if the user specifies `-debug` or `-verbose`, they will no longer get terminal feedback for piped commands, since `stdout.close()` will wait until the preceding command has completed. To do it properly, you'd probably have to instead handle the possibility of a broken pipe later on in `run.command()`, both during the `stderr.read(1)` loop and at the `process.wait()` call.

**Comment by jdtournier** [comment]
I fully agree, there's a problem with fe09164. The reason for this change was to follow the [advice on the official python docs](https://docs.python.org/2/library/subprocess.html#replacing-shell-pipeline) about how to handle broken pipes. But the current commit will cause the pipeline to run sequentially, by waiting for each process's `stdout` to close. The advice on the python docs seems to suggest that we should close each process's `stdout` in order of invocation, but _after_ all processes have been launched, so the better place to add these `stdout.close()` calls would be [here](https://github.com/MRtrix3/mrtrix3/blob/fe0916429a3c6832ef4cc91b6da2b5f74c75a6cf/lib/mrtrix3/run.py#L155) or thereabouts, and in the previous loop for verbose output, like @Lestropie says.

**Comment by maxpietsch** [comment]
The python doc only mentions the case of a single pipe. b5a04f7989f952e2 should hopefully fix it. I have tried to break it using multiple piped commands with broken pipes at various positions and it seems to work. Who thought that running a command via Python can be so finicky...

**Comment by Lestropie** [comment]
[This](https://github.com/MRtrix3/mrtrix3/commit/b5a04f7989f952e27fa5b34c2c16f247415d85d6#diff-15ea8e46c5e64749b52b4dff228aec59R153) will catch pipes that are broken once both processes are started; but what about if one of the processes aborts partway through? Is that a likely occurrence? If so, you'd need to catch it in the `while` loop.

> Who thought that running a command via Python can be so finicky...

Preaching to the choir... :-P

**Comment by maxpietsch** [comment]
> but what about if one of the processes aborts part way through

I think that this is the same problem as described [here](https://bugs.python.org/issue10963).

The more I google about it, the less I feel I understand how to do it... I think we basically reimplement `subprocess.communicate()` inside `run.command`. Might just as well use it?

**Comment by jdtournier** [comment]
> I think we basically reimplement subprocess.communicate() inside run.command. Might just as well use it?

Yes, it's starting to feel like it... @Lestropie, we were thinking that currently, all _MRtrix3_ pipelines are actually sequential: we just pass the filename of a temporary from one command to the next, and the first command won't write the filename to `stdout` until it's finished with it. So in essence, we could just as well run:

```ShellSession
$ tmp=$( dwi2tensor dwi.mif - )                  # catch stdout (the temporary filename) in variable tmp
$ tmp2=$( echo $tmp | tensor2metric - -fa - )    # pipe filename straight to next command
$ mrview $tmp2                                   # alternative syntax: supply tmp filename as argument 
```
This would be trivial to implement using _sequential_ invocations with `subprocess.communicate()`.

Only issue is this might get messy if/when we implement streamline piping - these types of jobs really would need to run concurrently...

**Comment by Lestropie** [comment]
> Only issue is this might get messy if/when we implement streamline piping

I'm expecting that I will want that in the not-too-distant future, which is why I'm trying to steer away from it.

**Comment by jdtournier** [comment]
>> Only issue is this might get messy if/when we implement streamline piping
>
> I'm expecting that I will want that in the not-too-distant future, which is why I'm trying to steer away from it.

Thought you might say that... and I have to agree, we need to come up with a solution that does support proper piping.
